sars2seq/translate.py

```diff
@@ -71,3 +140,77 @@ def translate(seq, name=None):
     seq += 'N' * (3 - remainder if remainder else 0)
 
     return Seq(seq).translate()
+
+
+def translateSpike(seq):
+    """
+    Translate a Spike sequence, taking into account gaps introduced in the
+    nucleotide alignment. This means that the amino acid sequences do not
+    have to be re-aligned after translating, which avoids the introduction of
+    gaps at the amino acid level that may be different from gaps at the
+    nucleotide level.
+
+    @param seq: A C{str} nucelotide sequence.
+    @return: A translated C{str} amino acid sequence that retains the gaps.
+    """
+    sequence = ''
+    current = 0
+    while current + 3 <= len(seq):
+        codon = seq[current:current+3]
+        if codon in CODONS:
+            # This is a codon that corresponds to a normal amino acid.
+            sequence += CODONS[codon]
+            current += 3
+        elif codon not in CODONS and codon.count('-') == 0:
+            # This is a codon that contains ambiguities.
+            sequence += 'X'
+            current += 3
+        elif codon.count('-') == 1:
+            # This is a codon that contains one gap (e.g. AT-).
+            # Count how many gaps there are after the current codon.
+            c = 3
+            while seq[current + c] == '-':
+                c += 1
+            subsequentGaps = c - 3
+
+            # Find the next nucleotide after the gap.
+            nextNt = seq[current + subsequentGaps + 3]
+
+            newCodon = codon[0:2] + nextNt
+
+            sequence += CODONS.get(newCodon, 'X')
+
+            # Add the correct number of gaps after the amino acid.
+            totalGaps = subsequentGaps + 1
+            gapAA = totalGaps // 3
+            sequence += (gapAA * '-')
```

Comment:
The parens aren't needed - but maybe you like/prefer them?

Reply:
I like them.